Typing Chinese characters in widget.Entry and perform undo/redo crashes the app.

### Checklist

- [x] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [x] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

widget.Entry undo/redo is not compatiable with Chinese characters. Undo/Redo crashes the app when dealing with multiple Chinese characters.

### How to reproduce

1. Create a text entry
2. Type random Chinese characters (for example, 测试) in the text entry
3. Call Undo()/Redo() or use the corresponding hotkey
4. Observe that the text entry transits into error states

//to crash the app
1. Create a text entry
2. Type random Chinese characters (for example, 测试) in the text entry
3. Select all and delete
4. Perform Undo() twice (ctrl + Z)
5. Observe that fyne crashes

### Screenshots

![image](https://github.com/user-attachments/assets/b1929a0f-634b-423b-9b9e-e85389fade0a)
![image](https://github.com/user-attachments/assets/f4416bb1-cc73-4298-8faf-453e629e764e)
![image](https://github.com/user-attachments/assets/aff0e672-f0d0-4e48-861e-b7e82e9d2879)


### Example code

Create text entry widget, type in Chinese characters, press ctrl + z, ctrl + y to perform undo and redo.

### Fyne version

2.5.0

### Go compiler version

1.22

### Operating system and version

Windows 10

### Additional Information

At: home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/internal/driver/glfw/driver_desktop.go:171
panic: runtime error: slice bounds out of range [4:2]

goroutine 11 [running]:
fyne.io/fyne/v2/widget.(*entryModifyAction).sub(...)
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/widget/entry.go:2316
fyne.io/fyne/v2/widget.(*entryModifyAction).Undo(0xc000f55ea0?, {0xc0053aadbe?, 0x7ff6a4a23872?})
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/widget/entry.go:2297 +0x90
fyne.io/fyne/v2/widget.(*entryUndoStack).Undo(0xc000d7fc08?, {0xc0053aadbe?, 0x7f0000c000f55ed8?})
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/widget/entry.go:2378 +0x4c
fyne.io/fyne/v2/widget.(*Entry).Undo(0xc000d7fc08)
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/widget/entry.go:787 +0x49
fyne.io/fyne/v2/widget.(*Entry).registerShortcut.func1({0xc000d7fc58?, 0x7ff6a5349920?})
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/widget/entry.go:1160 +0x17
fyne.io/fyne/v2.(*ShortcutHandler).TypedShortcut(0xc000d7fc58, {0x7ff6a55c8580, 0x7ff6a5b67b80})
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/shortcut.go:21 +0x6e
fyne.io/fyne/v2/widget.(*Entry).TypedShortcut(0xc0002eca20?, {0x7ff6a55c8580?, 0x7ff6a5b67b80?})
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/widget/entry.go:966 +0x25
fyne.io/fyne/v2/internal/driver/glfw.(*window).triggersShortcut.func1()
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/internal/driver/glfw/window.go:912 +0x24
fyne.io/fyne/v2/internal/driver/common.(*Window).RunEventQueue(...)
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/internal/driver/common/window.go:35
created by fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).createWindow.func1 in goroutine 1
        home/Go/dependency/pkg/mod/fyne.io/fyne/v2@v2.5.0/internal/driver/glfw/window.go:997 +0x136
2024/07/18 00:06:27 exit status 2
exit status 1

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -7,6 +7,7 @@ import (
 	"strings"
 	"time"
 	"unicode"
+	"unicode/utf8"
 
 	"fyne.io/fyne/v2"
 	"fyne.io/fyne/v2/canvas"
@@ -790,7 +791,7 @@ func (e *Entry) Undo() {
 	}
 	pos := modify.Position
 	if modify.Delete {
-		pos += len([]rune(modify.Text))
+		pos += utf8.RuneCountInString(modify.Text)
 	}
 	e.propertyLock.Lock()
 	e.updateText(newText, false)
@@ -2313,7 +2314,7 @@ func (i *entryModifyAction) add(s string) string {
 // Deletes Text
 func (i *entryModifyAction) sub(s string) string {
 	runes := []rune(s)
-	return string(runes[:i.Position]) + string(runes[i.Position+len([]rune(i.Text)):])
+	return string(runes[:i.Position]) + string(runes[i.Position+utf8.RuneCountInString(i.Text):])
 }
 
 func (i *entryModifyAction) TryMerge(other entryMergeableUndoAction) bool {
@@ -2342,13 +2343,13 @@ func (i *entryModifyAction) TryMerge(other entryMergeableUndoAction) bool {
 		}
 
 		if i.Delete {
-			if i.Position == other.Position+len([]rune(other.Text)) {
+			if i.Position == other.Position+utf8.RuneCountInString(other.Text) {
 				i.Position = other.Position
 				i.Text = other.Text + i.Text
 				return true
 			}
 		} else {
-			if i.Position+len([]rune(i.Text)) == other.Position {
+			if i.Position+utf8.RuneCountInString(i.Text) == other.Position {
 				i.Text += other.Text
 				return true
 			}

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -7,7 +7,6 @@ import (
 	"strings"
 	"time"
 	"unicode"
-	"unicode/utf8"
 
 	"fyne.io/fyne/v2"
 	"fyne.io/fyne/v2/canvas"
@@ -553,7 +552,7 @@ func (e *Entry) setText(text string, fromBinding bool) {
 func (e *Entry) Append(text string) {
 	e.propertyLock.Lock()
 	provider := e.textProvider()
-	provider.insertAt(provider.len(), text)
+	provider.insertAt(provider.len(), []rune(text))
 	content := provider.String()
 	changed := e.updateText(content, false)
 	cb := e.OnChanged
@@ -791,7 +790,7 @@ func (e *Entry) Undo() {
 	}
 	pos := modify.Position
 	if modify.Delete {
-		pos += utf8.RuneCountInString(modify.Text)
+		pos += len(modify.Text)
 	}
 	e.propertyLock.Lock()
 	e.updateText(newText, false)
@@ -939,7 +938,7 @@ func (e *Entry) TypedRune(r rune) {
 	pos := e.cursorTextPos()
 
 	provider := e.textProvider()
-	provider.insertAt(pos, string(runes))
+	provider.insertAt(pos, runes)
 
 	content := provider.String()
 	e.updateText(content, false)
@@ -947,7 +946,7 @@ func (e *Entry) TypedRune(r rune) {
 
 	e.undoStack.MergeOrAdd(&entryModifyAction{
 		Position: pos,
-		Text:     string(runes),
+		Text:     runes,
 	})
 	e.propertyLock.Unlock()
 
@@ -1038,9 +1037,7 @@ func (e *Entry) eraseSelection() bool {
 		return false
 	}
 
-	erasedText := e.Text[posA:posB]
-
-	provider.deleteFromTo(posA, posB)
+	erasedText := provider.deleteFromTo(posA, posB)
 	e.CursorRow, e.CursorColumn = e.rowColFromTextPos(posA)
 	e.selectRow, e.selectColumn = e.CursorRow, e.CursorColumn
 	e.selecting = false
@@ -1112,11 +1109,11 @@ func (e *Entry) pasteFromClipboard(clipboard fyne.Clipboard) {
 	runes := []rune(text)
 	pos := e.cursorTextPos()
 	provider := e.textProvider()
-	provider.insertAt(pos, text)
+	provider.insertAt(pos, runes)
 
 	e.undoStack.Add(&entryModifyAction{
 		Position: pos,
-		Text:     text,
+		Text:     runes,
 	})
 	content := provider.String()
 	e.updateText(content, false)
@@ -1610,7 +1607,7 @@ func (e *Entry) typedKeyReturn(provider *RichText, multiLine bool) {
 		return
 	}
 	e.propertyLock.Lock()
-	s := "\n"
+	s := []rune("\n")
 	pos := e.cursorTextPos()
 	provider.insertAt(pos, s)
 	e.undoStack.MergeOrAdd(&entryModifyAction{
@@ -2286,7 +2283,7 @@ type entryModifyAction struct {
 	// Position represents the start position of Text
 	Position int
 	// Text is the text that is inserted or deleted at Position
-	Text string
+	Text []rune
 }
 
 func (i *entryModifyAction) Undo(s string) string {
@@ -2308,13 +2305,13 @@ func (i *entryModifyAction) Redo(s string) string {
 // Inserts Text
 func (i *entryModifyAction) add(s string) string {
 	runes := []rune(s)
-	return string(runes[:i.Position]) + i.Text + string(runes[i.Position:])
+	return string(runes[:i.Position]) + string(i.Text) + string(runes[i.Position:])
 }
 
 // Deletes Text
 func (i *entryModifyAction) sub(s string) string {
 	runes := []rune(s)
-	return string(runes[:i.Position]) + string(runes[i.Position+utf8.RuneCountInString(i.Text):])
+	return string(runes[:i.Position]) + string(runes[i.Position+len(i.Text):])
 }
 
 func (i *entryModifyAction) TryMerge(other entryMergeableUndoAction) bool {
@@ -2325,7 +2322,7 @@ func (i *entryModifyAction) TryMerge(other entryMergeableUndoAction) bool {
 		}
 
 		// Don't merge two separate words
-		wordSeparators := func(s string) (num int, onlyWordSeparators bool) {
+		wordSeparators := func(s []rune) (num int, onlyWordSeparators bool) {
 			onlyWordSeparators = true
 			for _, r := range s {
 				if isWordSeparator(r) {
@@ -2343,14 +2340,14 @@ func (i *entryModifyAction) TryMerge(other entryMergeableUndoAction) bool {
 		}
 
 		if i.Delete {
-			if i.Position == other.Position+utf8.RuneCountInString(other.Text) {
+			if i.Position == other.Position+len(other.Text) {
 				i.Position = other.Position
-				i.Text = other.Text + i.Text
+				i.Text = append(other.Text, i.Text...)
 				return true
 			}
 		} else {
-			if i.Position+utf8.RuneCountInString(i.Text) == other.Position {
-				i.Text += other.Text
+			if i.Position+len(i.Text) == other.Position {
+				i.Text = append(i.Text, other.Text...)
 				return true
 			}
 		}

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -702,8 +702,7 @@ func (e *Entry) TypedKey(key *fyne.KeyEvent) {
 
 		e.propertyLock.Lock()
 		pos := e.cursorTextPos()
-		deletedText := e.Text[pos-1 : pos]
-		provider.deleteFromTo(pos-1, pos)
+		deletedText := provider.deleteFromTo(pos-1, pos)
 		e.CursorRow, e.CursorColumn = e.rowColFromTextPos(pos - 1)
 		e.undoStack.MergeOrAdd(&entryModifyAction{
 			Delete:   true,
@@ -718,8 +717,7 @@ func (e *Entry) TypedKey(key *fyne.KeyEvent) {
 		}
 
 		e.propertyLock.Lock()
-		deletedText := e.Text[pos : pos+1]
-		provider.deleteFromTo(pos, pos+1)
+		deletedText := provider.deleteFromTo(pos, pos+1)
 		e.undoStack.MergeOrAdd(&entryModifyAction{
 			Delete:   true,
 			Position: pos,
@@ -792,7 +790,7 @@ func (e *Entry) Undo() {
 	}
 	pos := modify.Position
 	if modify.Delete {
-		pos += len(modify.Text)
+		pos += len([]rune(modify.Text))
 	}
 	e.propertyLock.Lock()
 	e.updateText(newText, false)
@@ -2308,12 +2306,14 @@ func (i *entryModifyAction) Redo(s string) string {
 
 // Inserts Text
 func (i *entryModifyAction) add(s string) string {
-	return s[:i.Position] + i.Text + s[i.Position:]
+	runes := []rune(s)
+	return string(runes[:i.Position]) + i.Text + string(runes[i.Position:])
 }
 
 // Deletes Text
 func (i *entryModifyAction) sub(s string) string {
-	return s[:i.Position] + s[i.Position+len(i.Text):]
+	runes := []rune(s)
+	return string(runes[:i.Position]) + string(runes[i.Position+len([]rune(i.Text)):])
 }
 
 func (i *entryModifyAction) TryMerge(other entryMergeableUndoAction) bool {
@@ -2342,13 +2342,13 @@ func (i *entryModifyAction) TryMerge(other entryMergeableUndoAction) bool {
 		}
 
 		if i.Delete {
-			if i.Position == other.Position+len(other.Text) {
+			if i.Position == other.Position+len([]rune(other.Text)) {
 				i.Position = other.Position
 				i.Text = other.Text + i.Text
 				return true
 			}
 		} else {
-			if i.Position+len(i.Text) == other.Position {
+			if i.Position+len([]rune(i.Text)) == other.Position {
 				i.Text += other.Text
 				return true
 			}

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -556,6 +556,7 @@ func (e *Entry) Append(text string) {
 	content := provider.String()
 	changed := e.updateText(content, false)
 	cb := e.OnChanged
+	e.undoStack.Clear()
 	e.propertyLock.Unlock()
 
 	if changed {

--- a/widget/entry_test.go
+++ b/widget/entry_test.go
@@ -1984,7 +1984,84 @@ func TestEntry_CarriageReturn(t *testing.T) {
 	test.AssertImageMatches(t, "entry/carriage_return_text.png", w.Canvas().Capture())
 }
 
-func TestEntry_UndoRedo(t *testing.T) {
+func TestEntry_UndoRedo_TypeRune(t *testing.T) {
+	entry := widget.NewEntry()
+
+	// Check undo when there is nothing to undo
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "", entry.Text)
+
+	for _, r := range "aaa ààà bbb" {
+		entry.TypedRune(r)
+	}
+
+	assert.Equal(t, "aaa ààà bbb", entry.Text)
+
+	// Check redo when there is nothing to redo
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "aaa ààà bbb", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "aaa ààà", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "aaa", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "", entry.Text)
+
+	// Check undo when there is nothing to undo
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "aaa", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "aaa ààà", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "aaa ààà bbb", entry.Text)
+
+	// Check redo when there is nothing to redo
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "aaa ààà bbb", entry.Text)
+}
+
+func TestEntry_UndoRedo_Delete(t *testing.T) {
+	entry := widget.NewEntry()
+
+	// Check Undo when there is nothing to undo
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "", entry.Text)
+
+	for _, r := range "àbcdéf" {
+		entry.TypedRune(r)
+	}
+	assert.Equal(t, "àbcdéf", entry.Text)
+
+	entry.TypedKey(&fyne.KeyEvent{Name: fyne.KeyLeft})
+	entry.TypedKey(&fyne.KeyEvent{Name: fyne.KeyLeft})
+	entry.TypedKey(&fyne.KeyEvent{Name: fyne.KeyBackspace})
+	entry.TypedKey(&fyne.KeyEvent{Name: fyne.KeyBackspace})
+	entry.TypedKey(&fyne.KeyEvent{Name: fyne.KeyDelete})
+
+	assert.Equal(t, "àbf", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "àbéf", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "àbcdéf", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "àbéf", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "àbf", entry.Text)
+}
+
+func TestEntry_UndoRedoImage(t *testing.T) {
 	e, window := setupImageTest(t, true)
 	window.Resize(fyne.NewSize(128, 128))
 	c := window.Canvas()

--- a/widget/entry_test.go
+++ b/widget/entry_test.go
@@ -1991,21 +1991,21 @@ func TestEntry_UndoRedo_TypeRune(t *testing.T) {
 	entry.TypedShortcut(&fyne.ShortcutUndo{})
 	assert.Equal(t, "", entry.Text)
 
-	for _, r := range "aaa ààà bbb" {
+	for _, r := range "abc éàè 123" {
 		entry.TypedRune(r)
 	}
 
-	assert.Equal(t, "aaa ààà bbb", entry.Text)
+	assert.Equal(t, "abc éàè 123", entry.Text)
 
 	// Check redo when there is nothing to redo
 	entry.TypedShortcut(&fyne.ShortcutRedo{})
-	assert.Equal(t, "aaa ààà bbb", entry.Text)
+	assert.Equal(t, "abc éàè 123", entry.Text)
 
 	entry.TypedShortcut(&fyne.ShortcutUndo{})
-	assert.Equal(t, "aaa ààà", entry.Text)
+	assert.Equal(t, "abc éàè", entry.Text)
 
 	entry.TypedShortcut(&fyne.ShortcutUndo{})
-	assert.Equal(t, "aaa", entry.Text)
+	assert.Equal(t, "abc", entry.Text)
 
 	entry.TypedShortcut(&fyne.ShortcutUndo{})
 	assert.Equal(t, "", entry.Text)
@@ -2015,17 +2015,17 @@ func TestEntry_UndoRedo_TypeRune(t *testing.T) {
 	assert.Equal(t, "", entry.Text)
 
 	entry.TypedShortcut(&fyne.ShortcutRedo{})
-	assert.Equal(t, "aaa", entry.Text)
+	assert.Equal(t, "abc", entry.Text)
 
 	entry.TypedShortcut(&fyne.ShortcutRedo{})
-	assert.Equal(t, "aaa ààà", entry.Text)
+	assert.Equal(t, "abc éàè", entry.Text)
 
 	entry.TypedShortcut(&fyne.ShortcutRedo{})
-	assert.Equal(t, "aaa ààà bbb", entry.Text)
+	assert.Equal(t, "abc éàè 123", entry.Text)
 
 	// Check redo when there is nothing to redo
 	entry.TypedShortcut(&fyne.ShortcutRedo{})
-	assert.Equal(t, "aaa ààà bbb", entry.Text)
+	assert.Equal(t, "abc éàè 123", entry.Text)
 }
 
 func TestEntry_UndoRedo_Delete(t *testing.T) {
@@ -2059,6 +2059,29 @@ func TestEntry_UndoRedo_Delete(t *testing.T) {
 
 	entry.TypedShortcut(&fyne.ShortcutRedo{})
 	assert.Equal(t, "àbf", entry.Text)
+}
+
+func TestEntry_UndoRedo_Replace(t *testing.T) {
+	entry := widget.NewEntry()
+
+	entry.SetText("àbcdéf")
+	typeKeys(entry, fyne.KeyRight, fyne.KeyRight, keyShiftLeftDown, fyne.KeyRight, fyne.KeyRight, keyShiftLeftUp)
+	assert.Equal(t, "cd", entry.SelectedText())
+
+	entry.TypedRune('z')
+	assert.Equal(t, "àbzéf", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "àbéf", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutUndo{})
+	assert.Equal(t, "àbcdéf", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "àbéf", entry.Text)
+
+	entry.TypedShortcut(&fyne.ShortcutRedo{})
+	assert.Equal(t, "àbzéf", entry.Text)
 }
 
 func TestEntry_UndoRedoImage(t *testing.T) {

--- a/widget/richtext.go
+++ b/widget/richtext.go
@@ -160,9 +160,13 @@ func (t *RichText) charMinSize(concealed bool, style fyne.TextStyle, textSize fl
 }
 
 // deleteFromTo removes the text between the specified positions
-func (t *RichText) deleteFromTo(lowBound int, highBound int) string {
+func (t *RichText) deleteFromTo(lowBound int, highBound int) []rune {
+	if lowBound >= highBound {
+		return []rune{}
+	}
+
 	start := 0
-	var ret []rune
+	ret := make([]rune, 0, highBound-lowBound)
 	deleting := false
 	var segs []RichTextSegment
 	for i, seg := range t.Segments {
@@ -181,10 +185,8 @@ func (t *RichText) deleteFromTo(lowBound int, highBound int) string {
 
 		startOff := int(math.Max(float64(lowBound-start), 0))
 		endOff := int(math.Min(float64(end), float64(highBound))) - start
-		deleted := make([]rune, endOff-startOff)
 		r := ([]rune)(seg.(*TextSegment).Text)
-		copy(deleted, r[startOff:endOff])
-		ret = append(ret, deleted...)
+		ret = append(ret, r[startOff:endOff]...)
 		r2 := append(r[:startOff], r[endOff:]...)
 		seg.(*TextSegment).Text = string(r2)
 		segs = append(segs, seg)
@@ -200,7 +202,7 @@ func (t *RichText) deleteFromTo(lowBound int, highBound int) string {
 	}
 	t.Segments = segs
 	t.Refresh()
-	return string(ret)
+	return ret
 }
 
 // cachedSegmentVisual returns a cached segment visual representation.
@@ -251,7 +253,7 @@ func (t *RichText) cleanVisualCache() {
 }
 
 // insertAt inserts the text at the specified position
-func (t *RichText) insertAt(pos int, runes string) {
+func (t *RichText) insertAt(pos int, runes []rune) {
 	index := 0
 	start := 0
 	var into *TextSegment
@@ -276,7 +278,7 @@ func (t *RichText) insertAt(pos int, runes string) {
 	if pos > len(r) { // safety in case position is out of bounds for the segment
 		pos = len(r)
 	}
-	r2 := append(r[:pos], append([]rune(runes), r[pos:]...)...)
+	r2 := append(r[:pos], append(runes, r[pos:]...)...)
 	into.Text = string(r2)
 	t.Segments[index] = into
 }

--- a/widget/richtext_test.go
+++ b/widget/richtext_test.go
@@ -178,7 +178,7 @@ func TestText_InsertAt(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			text := NewRichTextWithText(tt.fields.buffer)
-			text.insertAt(tt.args.pos, tt.args.runes)
+			text.insertAt(tt.args.pos, []rune(tt.args.runes))
 			assert.Equal(t, tt.wantBuffer, text.String())
 		})
 	}
@@ -186,11 +186,11 @@ func TestText_InsertAt(t *testing.T) {
 
 func TestText_Insert(t *testing.T) {
 	text := NewRichTextWithText("")
-	text.insertAt(0, "a")
+	text.insertAt(0, []rune("a"))
 	assert.Equal(t, "a", text.String())
-	text.insertAt(1, "\n")
+	text.insertAt(1, []rune("\n"))
 	assert.Equal(t, "a\n", text.String())
-	text.insertAt(2, "b")
+	text.insertAt(2, []rune("b"))
 	assert.Equal(t, "a\nb", text.String())
 }
 
@@ -244,7 +244,7 @@ func TestText_DeleteFromTo(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			text := NewRichTextWithText(tt.fields.buffer)
 			got := text.deleteFromTo(tt.args.lowBound, tt.args.highBound)
-			assert.Equal(t, tt.want, got)
+			assert.Equal(t, tt.want, string(got))
 			assert.Equal(t, tt.wantBuffer, text.String())
 		})
 	}
@@ -332,7 +332,7 @@ func TestText_DeleteFromTo_Segments(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			text := NewRichText(tt.segments...)
 			got := text.deleteFromTo(tt.args.lowBound, tt.args.highBound)
-			assert.Equal(t, tt.want, got)
+			assert.Equal(t, tt.want, string(got))
 			for _, s := range tt.wantSegments {
 				if txt, ok := s.(*TextSegment); ok {
 					txt.parent = text

